File: Longest_Common_Substring.java
/*package whatever //do not write package name here */

import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
	public static void main (String[] args) {
		//code
		Scanner sc = new Scanner(System.in);
		
		int t = sc.nextInt();
		
		int ans = 0;
		
		for(int i = 0; i < t; i++){
		    ans = 0;
		    int n = sc.nextInt();
		    int m = sc.nextInt();
		    String x = sc.next();
		    String y = sc.next();
		    String temp = "";
		    for(int j = 0; j < n; j++){
		        for(int k = j; k < n; k++){
		            temp = temp + x.charAt(k);
		            int c = temp.length();
		            if((m-c)==0){
		                if(temp.equals(y)){
		                    ans = Math.max(ans, c);
		                }
		            }else{
    		            for(int l = 0; l < m-c; l++){
    		                String v = y.substring(l, l+c);
    		                if(v.equals(temp)){
    		                   
    		                    ans = Math.max(ans, c);
    		                    //System.out.println(ans);
    		                    break;
    		                }
    		            }
		            }
		        }
		        temp = "";
		    }
		    System.out.println(ans);
		}
	}
}

File: Longest_Common_Substring.py
for _ in range(int(input())):
    
    n,m = map(int, input().split())
    
    x = input()
    
    y = input()
    
    dpArr = [[0 for j in range(m+1)] for i in range(n+1)]
    
    ans = 0
    
    for i in range(1,n+1):
        
        for j in range(1,m+1):
            
            if x[i-1] == y[j-1]:
                
                dpArr[i][j] = 1 + dpArr[i-1][j-1]
                
                ans = max(ans, dpArr[i][j])
                
            else:
                
                dpArr[i][j] = 0
                
    print(ans)
